### PR TITLE
fix: compactor simulator math error creates 0 byte files

### PR DESCRIPTION
**Pull request by Joe-Blount**
This is a simulator (test) only change.
The simulator made a math error (a pair of matching off-by-1 mistakes) that causes the file bytes to incorrectly allocated in the simulations.  It is most noticeable on tiny files.  

As soon as this merges, I'll put up another little PR with a functional change to the compactor to support splitting files covering 2ns (fixing that issue got pretty crazy in the simulations without this PR, since it put 100% of the bytes one output file and 0 bytes in the other output file).

helps https://github.com/influxdata/idpe/issues/17974   (indirectly, as this and my next PR are required for the next step in the real work for that issue, which will be coming in a PR soon)

The first commit is the fix.  I suggest scrutinizing the fist commit.  
The second commit is all the insta churn resulting from it, which tweaks file sizes in many of our test cases because even the file sizes weren't crazy, they still were off a little in almost all cases.  Notably many of the 0b files are now a small, nonzero, size.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/tests/layouts/large_files.rs [hunk 1]
```diff
@@ -305,31 +305,33 @@ async fn two_large_files_total_over_max_compact_size() {
...
     - "**** Final Output Files (550mb written)"
     - "L1                                                                                                                 "
     - "L1.8[669,1000] 9ns 50mb                                                              |-----------L1.8------------| "
     - "L2                                                                                                                 "
     - "L2.6[669,1000] 8ns 50mb                                                              |-----------L2.6------------| "
-    - "L2.9[1,335] 9ns 100mb    |------------L2.9------------|                                                            "
-    - "L2.10[336,668] 9ns 100mb                               |-----------L2.10-----------|                               "
+    - "L2.9[1,334] 9ns 100mb    |------------L2.9------------|                                                            "
+    - "L2.10[335,667] 9ns 100mb                               |-----------L2.10-----------|                               "
+    - "L2.11[668,668] 9ns 307kb                                                             |L2.11|                       "
```
Comment (wiedld) [inline compactor/tests/layouts/large_files.rs:334]: An extra <1mb L2?

File: compactor/tests/layouts/many_files.rs [hunk 2]
```diff
@@ -4670,17 +4670,17 @@ async fn l0s_almost_needing_vertical_split() {
     - "L0.998[24,100] 1.02us    |-----------------------------------------L0.998-----------------------------------------|"
     - "L0.999[24,100] 1.02us    |-----------------------------------------L0.999-----------------------------------------|"
     - "L0.1000[24,100] 1.02us   |----------------------------------------L0.1000-----------------------------------------|"
-    - "**** Final Output Files (5.23gb written)"
+    - "**** Final Output Files (5.13gb written)"
+    - "L1                                                                                                                 "
+    - "L1.3091[92,100] 1.02us 88mb                                                                                |L1.3091| "
```
Comment (wiedld) [inline compactor/tests/layouts/many_files.rs:4675]: This now has an L1 in the output, whereas it didn't before?

File: compactor/tests/layouts/stuck.rs [hunk 3]
```diff
@@ -11395,126 +11410,119 @@ async fn stuck_l0_large_l0s() {
...
-    - "L2.3506[623,774] 199ns 100mb|L2.3506|                                                                                 "
-    - "L2.3507[775,925] 199ns 100mb|L2.3507|                                                                                 "
-    - "L2.3508[926,1072] 199ns 98mb|L2.3508|                                                                                 "
-    - "L2.3509[1073,1248] 199ns 100mb|L2.3509|                                                                                 "
-    - "L2.3510[1249,1423] 199ns 100mb|L2.3510|                                                                                 "
-    - "L2.3511[1424,1593] 199ns 98mb|L2.3511|                                                                                 "
-    - "L2.3512[1594,1703] 199ns 100mb|L2.3512|                                                                                 "
-    - "L2.3513[1704,1811] 199ns 99mb|L2.3513|                                                                                 "
-    - "L2.3514[2717,1990000] 199ns 2kb|----------------------------------------L2.3514----------------------------------------| "
```
Comment (wiedld) [inline compactor/tests/layouts/stuck.rs:11516]: Why was the previous one showing a large file (spatial representation) which is only 2kb?

Comment (wiedld) [reply compactor/tests/layouts/stuck.rs:11516]: And then with this change, we now (correctly) break it up and show the proper size?

**Review by wiedld — COMMENTED** [review]
The off-by-1 correction is small, but the impact on test output was much more than anticipated. Can we have a chat to understand a bit more about the test simulator?

File: compactor/tests/layouts/backfill.rs [hunk 4]
```diff
@@ -249,164 +249,208 @@ async fn random_backfill_empty_partition() {
...
+    - "**** Final Output Files (2.89gb written)"
     - "L2                                                                                                                 "
-    - "L2.86[0,208] 1.05us 100mb|-----L2.86------|                                                                        "
-    - "L2.92[209,406] 1.05us 100mb                   |-----L2.92-----|                                                      "
-    - "L2.93[407,603] 1.05us 100mb                                     |-----L2.93-----|                                    "
-    - "L2.94[604,701] 1.05us 50mb                                                       |L2.94-|                           "
-    - "L2.95[702,881] 1.05us 100mb                                                                |----L2.95-----|          "
-    - "L2.96[882,986] 1.05us 59mb                                                                                |-L2.96-| "
+    - "L2.96[0,220] 1.05us 100mb|------L2.96-------|                                                                      "
+    - "L2.100[644,818] 1.05us 100mb                                                          |---L2.100----|                 "
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/tests/layouts/backfill.rs:448]: we chatted in zoom, TLDR: changing the simulators math on estimating file sizes can change compaction decisions (e.g. how many files to compact in a round), which can affect the final file sizes & levels.  So long as both before & after the PR the results logical, legal, & we don't see a significant worsening in write amplification, its generally ok.

File: compactor/tests/layouts/large_files.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor/tests/layouts/large_files.rs:334]: (same comment as above)

File: compactor/tests/layouts/many_files.rs (hunk 2 above)
Comment (Joe-Blount) [reply compactor/tests/layouts/many_files.rs:4675]: (same comment as above)

File: compactor/tests/layouts/stuck.rs [hunk 5]
```diff
@@ -7389,159 +7389,159 @@ async fn stuck_l0_large_l0s() {
     - "L0                                                                                                                 "
     - "L0.?[1,906] 0ns 45mb     |-----------------L0.?-----------------|                                                  "
     - "L0.?[907,1811] 0ns 45mb                                          |-----------------L0.?-----------------|          "
-    - "L0.?[1812,2000] 0ns 10mb                                                                                  |-L0.?-| "
+    - "L0.?[1812,2000] 0ns 9mb                                                                                   |-L0.?-| "
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:7392]: Note that 9mb is split off from each of the 100MB input files, starting at 1812ns.  (here + the following 19 split operations)

File: compactor/tests/layouts/stuck.rs [hunk 6]
```diff
@@ -8472,8 +8472,8 @@ async fn stuck_l0_large_l0s() {
     - "L0.2285[1812,2000] 17ns  |----------------------------------------L0.2285-----------------------------------------|"
     - "L0.2288[1812,2000] 18ns  |----------------------------------------L0.2288-----------------------------------------|"
     - "L0.2291[1812,2000] 19ns  |----------------------------------------L0.2291-----------------------------------------|"
-    - "**** 1 Output Files (parquet_file_id not yet assigned), 190mb total:"
-    - "L0, all files 190mb                                                                                                "
+    - "**** 1 Output Files (parquet_file_id not yet assigned), 189mb total:"
+    - "L0, all files 189mb                                                                                                "
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:8476]: note that the 20 L0 files of 9mb split off above are compacted here to a single L0 of 189mb.

File: compactor/tests/layouts/stuck.rs [hunk 7]
```diff
@@ -9459,14 +9459,14 @@ async fn stuck_l0_large_l0s() {
...
-    - "L0.2905[1812,2000] 19ns 190mb|----L0.2905-----|                                                                        "
+    - "L0.2905[1812,2000] 19ns 189mb|----L0.2905-----|                                                                        "
     - "L0.2906[1812,2716] 172ns 0b|----------------------------------------L0.2906-----------------------------------------|"
     - "L0.2907[1812,2716] 192ns 0b|----------------------------------------L0.2907-----------------------------------------|"
     - "L0.2908[1812,2716] 199ns 0b|----------------------------------------L0.2908-----------------------------------------|"
-    - "**** 1 Output Files (parquet_file_id not yet assigned), 190mb total:"
-    - "L0, all files 190mb                                                                                                "
+    - "**** 1 Output Files (parquet_file_id not yet assigned), 189mb total:"
+    - "L0, all files 189mb                                                                                                "
     - "L0.?[1812,2716] 199ns    |------------------------------------------L0.?------------------------------------------|"
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:9470]: Here the 189mb is compacted with L0s extending to 2716ns.  Even though those other L0s were 0b, the simulator will now assume the 189mb is spread evenly across 1812-2716ns.

File: compactor/tests/layouts/stuck.rs [hunk 8]
```diff
@@ -11395,126 +11410,119 @@ async fn stuck_l0_large_l0s() {
...
+    - "L1                                                                                                                 "
+    - "L1.3500[2717,1990000] 199ns 2kb|----------------------------------------L1.3500----------------------------------------| "
+    - "L1.3498[1682,1811] 199ns 87mb|L1.3498|                                                                                 "
+    - "L1.2953[1812,2716] 199ns 189mb|L1.2953|                                                                                 "
+    - "**** 3 Output Files (parquet_file_id not yet assigned), 276mb total:"
+    - "L2                                                                                                                 "
+    - "L2.?[1682,723271] 199ns 100mb|-------------L2.?-------------|                                                          "
+    - "L2.?[723272,1444860] 199ns 100mb                                |-------------L2.?-------------|                          "
+    - "L2.?[1444861,1990000] 199ns 76mb                                                                 |---------L2.?---------| "
```
Comment (Joe-Blount) [inline compactor/tests/layouts/stuck.rs:11496]: The 189mb noted from above is now compacted with the 2kb that covers out to 1,990,000ns, and another file of 87mb.  The simulator assumes data is spread linearly, so it assumes the output 176mb is spread evenly across the entire range from 1682-1,990,000ns.

Even though that doesn't fit the input files we started with, with the steps above and the simulators assumed linear distribution of output, this type of thing is to be expected.  It is simulator quirk, not an issue in the real compactor code.

File: compactor/tests/layouts/stuck.rs (hunk 3 above)
Comment (wiedld) [reply compactor/tests/layouts/stuck.rs:11516]: Confirm the total mb does not significantly changed, and that the compactor simulation is moving around simulated bytes (based on what you highlighted above and in comments below). Thank you for doing the extra work to confirm we are ok.

**Review by wiedld — APPROVED** [review]

**Pull request merged**